pkg/net/libp2p/libp2ptransport.go

```diff
@@ -0,0 +1,283 @@
+	return nil
+}
+
+func (t *Transport) Stop() {
+	t.logger.Log(logging.LevelDebug, "Stopping libp2p transport.")
+	defer t.logger.Log(logging.LevelDebug, "libp2p transport stopped.")
+
+	t.outboundStreamsMx.Lock()
+	defer t.outboundStreamsMx.Unlock()
+
+	for id, s := range t.outboundStreams {
+		if s == nil {
+			continue
+		}
+		t.logger.Log(logging.LevelDebug, "Closing connection", "to", id)
+		if err := s.Close(); err != nil {
+			t.logger.Log(logging.LevelWarn, fmt.Sprintf("Could not close connection to node %v: %v", id, err))
+			continue
+		}
+		t.logger.Log(logging.LevelDebug, "Closed connection", "to", id)
+	}
+
+	if err := t.host.Close(); err != nil {
+		t.logger.Log(logging.LevelError, fmt.Sprintf("Could not close libp2p %v: %v", t.ownID, err))
+	} else {
+		t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %v libp2p host closed", t.ownID))
+	}
+}
+
+func (t *Transport) Connect(ctx context.Context) {
+	wg := sync.WaitGroup{}
+	wg.Add(len(t.membership) - 1)
+
+	for nodeID, nodeAddr := range t.membership {
+		if nodeID == t.ownID {
+			continue
+		}
+
+		go func(nodeID types.NodeID, nodeAddr string) {
+			defer wg.Done()
+
+			t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %s is connecting to node %s", t.ownID.Pb(), nodeID.Pb()))
+
+			maddr, err := multiaddr.NewMultiaddr(nodeAddr)
+			if err != nil {
+				t.logger.Log(logging.LevelError,
+					fmt.Sprintf("failed to parse %v for node %v: %v", nodeAddr, nodeID, err))
+				return
+			}
+
+			// Extract the peer ID from the multiaddr.
+			info, err := peer.AddrInfoFromP2pAddr(maddr)
+			if err != nil {
+				t.logger.Log(logging.LevelError,
+					fmt.Sprintf("failed to parse addr %v for node %v: %v", maddr, nodeID, err))
+				return
+			}
+
+			t.host.Peerstore().AddAddrs(info.ID, info.Addrs, peerstore.PermanentAddrTTL)
+
+			s, err := t.openStream(ctx, info.ID)
+			if err != nil {
+				t.logger.Log(logging.LevelError, fmt.Sprintf("couldn't open stream: %v", err))
+				return
+			}
+			t.addOutboundStream(nodeID, s)
+			t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %s has connected to node %s", t.ownID.Pb(), nodeID.Pb()))
+
+		}(nodeID, nodeAddr)
+	}
+
+	wg.Wait()
+}
+
+func (t *Transport) openStream(ctx context.Context, p peer.ID) (network.Stream, error) {
+	timeout := 500 * time.Millisecond
+	for {
+		sctx, cancel := context.WithTimeout(ctx, timeout)
+		s, err := t.host.NewStream(sctx, p, ID)
+		cancel()
+		if err != nil {
+			t.logger.Log(logging.LevelDebug, "opening stream", "err", err.Error())
+			select {
+			case <-time.After(timeout):
+				continue
+			case <-ctx.Done():
+				return nil, fmt.Errorf("context closed")
+			}
+		}
+		return s, nil
+	}
+}
+
+func (t *Transport) Send(dest types.NodeID, payload *messagepb.Message) error {
+	outBytes, err := proto.Marshal(payload)
+	if err != nil {
+		return fmt.Errorf("failed to marshal message: %w", err)
+	}
+
+	t.outboundStreamsMx.Lock()
+	s, ok := t.outboundStreams[dest]
+	t.outboundStreamsMx.Unlock()
```

Comment:
If we expect several routines to write at the same time to the stream (i.e. they are able to send messages in parallel) we should defer a bit the release of the lock. Writers are not thread-safe: https://stackoverflow.com/questions/19646717/is-the-go-bytes-buffer-thread-safe

Reply:
In this implementation, sending happens sequentially. `addOutboundStream` seems to also access this map, so the lock might be necessary. In general, this is not the most efficient implementation, as sending to different destinations could be done in parallel, but this implementation is simple and will do for the moment.